### PR TITLE
[SDP-907] data/db: Automatic cancellation of Ready Payments after a certain time period

### PR DESCRIPTION
**Pull request by ceciliaromao**
### What

Automatic cancellation of pending payments after a certain time period

### Why

If a payment has been in 'ready' status for more than a certain quantity of days, move from 'ready' to 'canceled' status. Do not pay the receiver if they verify their identity after that.

### Known limitations

N/A

#### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).
* [x] This PR's title starts with the name of the package, area, or subject affected by the change.

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [ ] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [x] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

**Review by CaioTeixeira95 — COMMENTED** [review]
Thank you for working on this! I've left some comments. Let me know if you have any questions.

File: internal/data/payments.go [hunk 1]
```diff
@@ -640,3 +640,27 @@ func newPaymentQuery(baseQuery string, queryParams *QueryParams, paginated bool,
...
 }
+
+// CancelPayments cancels automatically payments that are in "READY" status for an informed time period in days.
+func (p *PaymentModel) CancelPayments(ctx context.Context, sqlExec db.SQLExecuter, periodInDays int64) error {
+	query := `
+		UPDATE payments
+		SET status = 'CANCELED'::payment_status,
+				status_history = array_append(status_history, create_payment_status_history(NOW(), 'CANCELED', NULL))
+		WHERE status = 'READY'::payment_status
+		AND updated_at <= $1
```
Comment (marwen-abid) [inline internal/data/payments.go:651]: Please correct if I'm wrong, but I don't believe `updated_at` is a good measure here for cancellation period of a payment. 

Updating any field in a payment can trigger `updated_at` to be changed. 

The only field that seems appropriate is the status history timestamp of the last ready status. e.g. 
`{\"status\": \"READY\", \"timestamp\": \"2023-10-16T21:50:47.911779+00:00\", \"status_message\": null}`

File: internal/db/migrations/2023-10-25-alter-organizations-table.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+-- +migrate Up
+ALTER TABLE
+  public.organizations
+ADD
+  COLUMN payment_cancellation_period int;
```
[the remark this answers is not in the document]
Comment (marwen-abid) [reply internal/db/migrations/2023-10-25-alter-organizations-table.sql:5]: If payment cancellation period unit is always days, we should change this to `payment_cancellation_period_days`

File: internal/services/ready_payments_cancellation_service.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		return fmt.Errorf("error getting organization: %w", err)
+	}
+
+	if organization.PaymentCancellationPeriod == nil {
+		log.Debug("automatic ready payment cancellation is deactivated. Set a valid value to the organization's payment_cancellation_period to activate it.")
+		return nil
+	}
+
+	return db.RunInTransaction(ctx, s.sdpModels.DBConnectionPool, nil, func(dbTx db.DBTransaction) error {
```
Comment (marwen-abid) [inline internal/services/ready_payments_cancellation_service.go:40]: Does this need to run in a transaction if it's only executing one update statement ?

File: internal/data/payments.go [hunk 4]
```diff
@@ -640,3 +640,27 @@ func newPaymentQuery(baseQuery string, queryParams *QueryParams, paginated bool,
 	query, params := qb.Build()
 	return sqlExec.Rebind(query), params
 }
+
+// CancelPayments cancels automatically payments that are in "READY" status for an informed time period in days.
+func (p *PaymentModel) CancelPayments(ctx context.Context, sqlExec db.SQLExecuter, periodInDays int64) error {
```
[the remark this answers is not in the document]
Comment (ceciliaromao) [reply internal/data/payments.go:645]: Good ideia 🚀

File: internal/data/payments_test.go [hunk 5]
```diff
@@ -1256,3 +1257,267 @@ func Test_PaymentModelGetAllReadyToPatchCompletionAnchorTransactions(t *testing.
...
+		require.NoError(t, err)
+
+		payment2DB, err := models.Payment.Get(ctx, payment2.ID, dbConnectionPool)
+		require.NoError(t, err)
+
+		assert.Equal(t, DraftPaymentStatus, payment1DB.Status)
+		assert.Equal(t, ReadyPaymentStatus, payment2DB.Status)
+	})
+
+	t.Run("multiple ready cases", func(t *testing.T) {
```
Comment (CaioTeixeira95) [inline internal/data/payments_test.go:1354]: Could you please improve this test description? Perhaps `successfully cancel payments when it has multiple ready status entries`

```suggestion
	t.Run("successfully cancel payments when it has multiple ready status history entries", func(t *testing.T) {
```

File: internal/data/payments.go [hunk 6]
```diff
@@ -640,3 +640,37 @@ func newPaymentQuery(baseQuery string, queryParams *QueryParams, paginated bool,
...
+			status_history = array_append(status_history, create_payment_status_history(NOW(), 'CANCELED', NULL))
+		WHERE 
+			status = 'READY'::payment_status
+			AND (
+				SELECT (value->>'timestamp')::timestamp
+				FROM unnest(status_history) AS value
+				WHERE value->>'status' = 'READY' 
+				ORDER BY (value->>'timestamp')::timestamp DESC 
+				LIMIT 1
+			) <= $1
```
Comment (CaioTeixeira95) [inline internal/data/payments.go:660]: 🚀

File: internal/serve/httphandler/profile_handler.go [hunk 7]
```diff
@@ -49,6 +49,7 @@ type PatchOrganizationProfileRequest struct {
 	TimezoneUTCOffset              string  `json:"timezone_utc_offset"`
 	IsApprovalRequired             *bool   `json:"is_approval_required"`
 	SMSResendInterval              *int64  `json:"sms_resend_interval"`
+	PaymentCancellationPeriod      *int64  `json:"payment_cancellation_period_days"`
```
Comment (CaioTeixeira95) [inline internal/serve/httphandler/profile_handler.go:52]: You've changed the json tag name but not the struct attribute.

```suggestion
	PaymentCancellationPeriodDays      *int64  `json:"payment_cancellation_period_days"`
```

**Review by CaioTeixeira95 — APPROVED** [review]
Awesome work! Thank you for working on this! 🚀

**Pull request merged**